ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -445,10 +445,17 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
       obsoleteDirs = dir.getAbortedDirectories();
     }
 
-    if (obsoleteDirs.isEmpty()
-        && !hasDataBelowWatermark(dir, fs, path, ci.highestWriteId, writeIdList.getHighWatermark())) {
-      LOG.info(idWatermark(ci) + " nothing to remove below watermark " + ci.highestWriteId + ", ");
-      return true;
+    if (obsoleteDirs.isEmpty()) {
+      conf.set(ValidTxnList.VALID_TXNS_KEY, new ValidReadTxnList().toString());
+      
+      AcidDirectory fullDir = AcidUtils.getAcidState(fs, path, conf,
+        new ValidReaderWriteIdList(ci.getFullTableName(), new long[0], new BitSet(), ci.highestWriteId, Long.MAX_VALUE),
+        Ref.from(false), false, hdfsDirSnapshots);
+      
+      if (fullDir.getObsolete().isEmpty()) {
+        LOG.info(idWatermark(ci) + " nothing to remove below watermark " + ci.highestWriteId + ", ");
+        return true;
+      }
```

Comment:
It might be a bit more readable/simpler to add an else clause (handle the case of !obsoleteDirs.isEmpty()) here, what do you think?

Reply:
refactored, please take a look.